src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentProjectsFinder.cs

```diff
@@ -128,110 +68,87 @@ private static ImmutableArray<Project> GetProjects(Solution solution, ImmutableA
-            if (containingAssembly == null)
-            {
-                // currently we don't support finding references for a symbol that doesn't have containing assembly symbol
-                return ImmutableArray<Project>.Empty;
-            }
-
-            // Find the projects that reference this assembly.
+            var symbolOrigination = GetSymbolOrigination(solution, symbol, cancellationToken);
 
-            // If this is a source symbol from a project, try to find that project.
-            var sourceProject = solution.GetProject(containingAssembly, cancellationToken);
-            cancellationToken.ThrowIfCancellationRequested();
+            // If we can't find where the symbol came from, we can't determine what projects to search for references to it.
+            if (symbolOrigination.assembly == null)
+                return ImmutableArray<Project>.Empty;
 
             // 1) Compute all the dependent projects (submission + non-submission) and their InternalsVisibleTo semantics to the definition project.
-            ImmutableArray<DependentProject> dependentProjects;
 
             var visibility = symbol.GetResultantVisibility();
             if (visibility == SymbolVisibility.Private)
             {
-                dependentProjects = await GetDependentProjectsCoreAsync(symbol, solution, sourceProject, visibility, cancellationToken).ConfigureAwait(false);
+                // For private symbols, we only need the current project (and related submissions).  No need to cache
+                // that, just simply compute and return the result.
+                return GetProjects(solution, await ComputeDependentProjectsAsync(solution, symbolOrigination, visibility, cancellationToken).ConfigureAwait(false));
             }
-            else
-            {
-                // We cache the dependent projects for non-private symbols, check in the cache first.
-                var dependentProjectsMap = s_dependentProjectsCache.GetValue(solution, s_createDependentProjectsMapCallback);
-                var key = new DefinitionProject(sourceProjectId: sourceProject?.Id, assemblyName: containingAssembly.Name.ToLower());
 
-                if (!dependentProjectsMap.TryGetValue(key, out dependentProjects))
-                {
-                    dependentProjects = await GetDependentProjectsCoreAsync(symbol, solution, sourceProject, visibility, cancellationToken).ConfigureAwait(false);
-                    dependentProjectsMap.TryAdd(key, dependentProjects);
-                }
-            }
+            // Otherwise, for non-private symbols, we cache the dependent projects for non-private symbols to speed up
+            // future calls.
+            var dependentProjectsMap = s_dependentProjectsCache.GetValue(solution, _ => new DependentProjectMap());
+
+            var asyncLazy = dependentProjectsMap.GetOrAdd(
+                new DefinitionProject(symbolOrigination.sourceProject?.Id, symbolOrigination.assembly.Name.ToLower()),
```

Comment:
Is the "name" here just the name or the full identity of the assembly w/ version and other stuff? Should this just be using IAssemblySymbol.Identity rather than the extra string allocations?

Reply:
What happens if two unrelated projects point to different versions of an assembly on disk that have different IVTs?

Reply:
1. the name is the simple name of the assembly, without all the identity goop for it.
2. if two unrelated projects point to different versions of the assembly, we'll likely say that both projects have a dependency on the assembly, nd we will consider them candidates for searching.
3. This would only really be relevant if somehow you had those multiple projects referencing multiple different assemblies *and* those assemblies have IVT to those projects specifically somehow.
4. even in that rare case, all that would happen is that we would just search that project for hits.

That's really all that's going on here.  This code is not about correctness, it's about performance.  i.e. it's trying to limit the number of projects we look at.  We could technically look at all projects, however that would be *much* slower since we'd have to make all the compilations ofr all of them.

So:

previously: we would strictly figure out the set of projects P_1 that we definitely needed to search.
now: we figure out the set of projects P_2 which is equal to or a superset of P_1.  We save all the time computing P_1, but we will pay extra time searching the extra `P_2 - P_1`.  In practice, i expect that extra set of projects to be empty, and even if there are a couple of extra projects in it, it should still be much faster than checking all the extra projects like we do today.

Reply:
My concern is this scenario:

1. Project A depends on a Metadata.dll that _doesn't_ have any IVT attributes in it.
2. Project B also depends on a different Metadata.dll, but it does use internals via IVT from Metadata.dll.
3. Some search happens on a public type in Metadata.dll, and the analysis is done with the IAssemblySymbol that is seen through A; the cache is populated that Metadata.dll is consumed from A and B, but because we're using the IAssemblySymbol of the Metadata.dll as seen through A, neither are marked with "has internal access".
4. A search happens in B's code on an internal type obtained from Metadata.dll through the IVT B sees. Because it's internal, we fetch the entry from step 3, and filter it to only things with internal access, which is nothing. B's code is never searched.

I see a few approaches for this:

1. Delete the cache. Problem solved; I think you said privately you are already doing that, so maybe that's the easiest answer.
2. [If we have to restore the cache...] Don't care. I'm hesitant to say this only because I don't know how widespread this is, I do know how difficult it will be to figure out a customer report if it does happen. I'm also wondering if Roslyn could hit this itself: our CodeStyle layer builds against NuGet packages that are older: would we end up building the IVT list based upon the IVTs in the NuGet package we build against, a newer IVT added in the source might potentially get skipped...? Consider my scenario above, where Project A is our Code Style layer and B is some code that has IVT into our Workspaces layer. Minimally, if we choose to not care, a comment at least recognizing the cache issue might be wise so that way if somebody else is debugging and thinks the cache is wrong, they know they're not crazy.
3. [If we can't delete the cache and find we must care...] Fix it. We already have to walk through all projects and all IAssemblySymbols to actually see which projects are referencing the assembly name in question; at that point we've collected all the IAssemblySymbols that might have an IVT we need to consider -- if we union all of the IVTs together of all the those assembly symbols, then the cache is safe. Assuming you don't process the same IAssemblySymbol more than once, this shouldn't really be any slower than what we have now. I'm guessing multitargeting might mean each IAssemblySymbol might be different, but it's easy to ask if two come from the same PE file path on disk, which would be sufficient for our needs here.

Reply:
> Delete the cache. Problem solved; I think you said privately you are already doing that, so maybe that's the easiest answer.

I'm going to take this approach.  In my branch i see no impact in removing it.  Do you want that in this PR, or in the followup?

Reply:
Let's just follow up since you already have that one readyish.